Eclipse fixes

dpursehouse opened the pull request:
Various minor fixes to make the project buildable in Eclipse.

dpursehouse committed:
Fix "Plugin execution not covered by lifecycle configuration" in Eclipse

When the project is imported into Eclipse, the build fails with:

  "Plugin execution not covered by lifecycle configuration"

for the maven-antrun-plugin.

According to [1] the solution is to enclose all plugins within the
"pluginManagement" tag in the POM file.

[1] http://stackoverflow.com/a/13733232/381622

Change-Id: I72fd0fd0b885df1100fc178f8cd70361daa83eab

dpursehouse committed:
Update Eclipse core prefs

Eclipse Luna overwrites the prefs file, putting all the settings
in alphabetical order and removing comments. This causes the prefs
files to be modified in the git workspace.

Update the file with the version generated by Eclipse to prevent it
from being modified every time.

No settings values are modified by this change.

Change-Id: I9e245b7614ed0bf34314c8c459da81204aab84bc

dpursehouse committed:
Set encoding to UTF-8 in Eclipse prefs

Change-Id: I0237d139c2ee8df04fb12901f2ee4484c2ac9e30

dpursehouse committed:
Fix "Error loading property file" in Eclipse

Prefix property file paths with "${basedir}/" as described in [1].

[1] http://stackoverflow.com/a/23481166/381622

Change-Id: I3a1b3d813c0358f6aa29c9f81e2913900f9c9728

jpountz commented on .settings/org.eclipse.jdt.core.prefs:
```diff
@@ -1,21 +1,18 @@
 org.eclipse.jdt.core.compiler.annotation.nullanalysis=enabled
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.7
+org.eclipse.jdt.core.compiler.compliance=1.7
+org.eclipse.jdt.core.compiler.problem.forbiddenReference=warning
```
Can you explain what it does?

dpursehouse replied:
Sorry, I didn't notice that this had been added.

According to the documentation:  "When enabled, the compiler will issue an error or a warning when referring to a type with restricted access, as defined according to the access restriction specifications."

The default value is "warning".

I guess the version of Eclipse that originally generated these prefs didn't have that option, while mine is later and does have it.

jpountz commented on .settings/org.eclipse.core.resources.prefs:
```diff
@@ -1,2 +1,7 @@
+encoding//src/test/java=UTF-8
+encoding//src/test/resources=UTF-8
 encoding/<project>=UTF-8
+encoding/rest-api-spec=UTF-8
```
If we specify the encoding at the project level, shouldn't all source folders inherit it?

jpountz commented on .settings/org.eclipse.jdt.core.prefs:
```diff
@@ -1,21 +1,18 @@
+org.eclipse.jdt.core.compiler.source=1.7
+org.eclipse.jdt.core.formatter.lineSplit=140
+org.eclipse.jdt.core.formatter.tabulation.char=space
+org.eclipse.jdt.core.formatter.tabulation.size=4
```
Something I liked with the previous layout is that I knew what these options did. It is a bit more cryptic now. You commit message mentions Eclipse Luna but I am using Luna too and I don't have the issue?

dpursehouse replied:
In all my projects the settings get listed in alphabetical order like this.  I haven't done anything in Eclipse to make it do that.

> it is a bit more cryptic now

The names of the settings make the meaning clear enough to me.  In any case, don't most people use the UI to change these rather than manually editing the file?

jpountz commented on pom.xml:
```diff
@@ -1541,9 +1542,7 @@
-        </plugins>
-        <pluginManagement>
-            <plugins>
+
```
This is the most tricky bit to me. I'd be wary about modifying maven configuration just for eclipse.

dpursehouse replied:
I couldn't find any other way to make it build in Eclipse.  If you're using Eclipse, how do you get rid of the error without changing this config?

jpountz replied:
I guess I don't have the error because I am not using the Maven integration plugin. I'll try to ping some other people who are using Eclipse to see how they do.

jpountz replied:
More specifically, I run mvn eclipse:eclipse and then import the project in Eclipse instead of expecting Eclipse to be able to understand the pom.xml through a plugin.

dpursehouse replied:
Yes, I'm using the Maven plugin in Eclipse and use that to generate the project.  I'll try it the way you suggest and see if that helps.

dpursehouse replied:
Generating the project with `mvn eclipse:eclipse` doesn't seem to have made any difference.  I still get the errors about the plugin execution not covered by lifecycle configuration, and errors loading the property files.

Not sure if that's because my Eclipse has the m2e plugins installed, though.  I don't want to remove those since I need them for other projects.

colings86 replied:
@dpursehouse after generating the eclipse files using eclipse:eclipse are you still importing the project using 'existing maven projects' in the Import dialog? You will need to instead use 'Existing Projects into Workspace' under the 'general' folder in that dialog.

dpursehouse replied:
@colings86 after completing nuking the project and starting again with `mvn eclipse:eclipse` and then importing as suggested, it works.

Thanks for the help, and sorry for the noise caused by this PR.

The other warning about duplicate declarations has since been fixed in e60f61d0a9e928bcd1ea6ee965e79f5003f76a95 so I will close this PR.

jpountz commented:
Thanks for proposing fixes @dpursehouse, I left some questions inline.

dpursehouse committed:
Fix duplicate declarations of maven-resources-plugin in pom.xml

Maven reports the following warning at the beginning of the build:

  'build.pluginManagement.plugins.plugin.(groupId:artifactId)' must be unique but found
  duplicate declaration of plugin org.apache.maven.plugins:maven-resources-plugin

  It is highly recommended to fix these problems because they threaten the stability
  of your build.

  For this reason, future Maven versions might no longer support building such malformed
  projects.

The maven-resources-plugin is declared three times, each with a different
<execution> section.

Consolidate the <execution> sections from the duplicate declarations under
the <executions> section in a single declatation of the plugin.

Change-Id: I16afcc0a36c118098b69379d182e1cd43df0d2be

dpursehouse committed:
Fix another instance of "Error loading property file" in Eclipse

Change-Id: I45c454789886a41ea0cf3b40f7dafe5dd5923d77

dpursehouse commented:
Note: I have signed the CLA (as an individual contributing under an existing company contributor agreement) and an email has been sent to get me added to the contributors list.

jpountz commented:
@colings86 @markharwood @rmuir I think you are using Eclipse, do you have similar issue as @dpursehouse (see in particular comments on the pom.xml file)

colings86 commented:
@dpursehouse instead of modifying the plugins in the pom.xml directly to remove the errors in the maven integration it is often better to use the eclipse lifecycle mapping plugin to tell eclipse how to deal with these plugins. you can see an example of this [here](https://github.com/elastic/elasticsearch/pull/10495/files#diff-600376dffeb79835ede4a0b285078036R1547) where we are telling eclipse to execute the copy-dependencies goal of the maven-dependancy-plugin as part of a full or incremental build. You can also instruct eclipse to ignore a plugin by using `<ignore/>` instead of `<execute/>`.

The two plugins that are erroring I believe are the maven-antrun-plugin and the maven-resources-plugin. Is that what you are seeing? The ant-run plugin does some validation so might be useful to execute as part of the eclipse build but the resources plugin is only relevant to the creation of the deb and rpm packages so can be ignored for the eclipse build.

Maybe you could update your PR to use this approach since it does not involve changing the pom for non-eclipse builds?

rmuir commented:
@jpountz eclipse is working fine for me, and has been for months.

colings86 commented:
@dpursehouse I ended up fixing this using some of your code is https://github.com/elastic/elasticsearch/commit/97a9b4ec2fb6691437c86b694b77d56e81efcaa1

This should mean that you can now import the project in eclipse using the 'import existing maven projects' option.

Thanks for raising this problem and for providing code for the fix.

This has been merged into master, 1.x, and 1.5

dpursehouse commented:
@colings86 Thanks.  It works perfectly now.
